### PR TITLE
The `APIClient` now operates on `URL` types instead of `Strings` wher…

### LINKED ISSUE #9: Use URL instead of String, where possible.

### PR DESCRIPTION
**Pull request by vinczebalazs**
…e possible. Fixes #9.

The only place where `Strings` had to be used is the `url` property of a `PageItem`. Here it is possible that the given page item is related to weather (e.g. City Weather) where the `url` contains placeholders. Placeholders are not supported by the built in `URL` type.

### CODE REVIEW

File: ApiClient/Source/SchedJoulesApiClient.swift [hunk 1]
```diff
@@ -1,5 +1,5 @@
 //
-//  SchedJoulesApiClient.swift
+//  API.swift
```
Comment (dmfs) [inline ApiClient/Source/SchedJoulesApiClient.swift]: Wrong file name ;-)

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 2]
```diff
@@ -27,25 +27,33 @@ import Foundation
...
+        self.init(path: "", queryItems: [localeQuery,locationQuery])
     }
     
     /// Manualy specify locale and location parameters
-    convenience init(locale: String, location: String){
-        self.init(resource: "pages/?locale=\(locale)?location=\(location)")
+    convenience init(locale: String, location: String) {
+        let localeQuery = URLQueryItem(name: "locale", value: locale)
+        let locationQuery = URLQueryItem(name: "location", value: location)
+        self.init(path: "", queryItems: [localeQuery,locationQuery])
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: Don't create these redundant local variables. This can be written much more readable like this
```swift
        self.init(path: "", queryItems: [
               URLQueryItem(name: "locale", value: locale),
               URLQueryItem(name: "location", value: location)])
```

File: ApiClient/Source/Pages/Queries/SearchQuery.swift [hunk 3]
```diff
@@ -27,22 +27,26 @@ import Foundation
...
+    let url: URL?
+    let host: String = "https://api.schedjoules.com/"
     let method: HTTPMethod = .get
     let parameters: Parameters = [:]
     let headers: HTTPHeaders = ["Accept" : "application/json", "Content-Type" : "application/json"]
     
-    required init(resource: String) {
-        self.resource = resource
+    required init(path: String, queryItems: [URLQueryItem]) {
+        var urlComponents = URLComponents(string: host)
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/SearchQuery.swift]: is it possible to write this like so?
```swift
self.url = URLComponents(string: host, path: path, queryItems: queryItems).url
```

File: ApiClient/Source/Pages/Queries/SearchQuery.swift [hunk 4]
```diff
@@ -27,22 +27,26 @@ import Foundation
...
+        urlComponents?.path = path
+        urlComponents?.queryItems = queryItems
+        self.url = urlComponents?.url
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/SearchQuery.swift]: why do you use the `?` operator? Can `urlComponents` be `null` at this line? If so, why can that be? What's the expected behavior in that case?

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift (hunk 2 above)
Comment (vinczebalazs) [reply ApiClient/Source/Pages/Queries/HomePageQuery.swift]: Fixing.

File: ApiClient/Source/Pages/Queries/SearchQuery.swift (hunk 3 above)
Comment (vinczebalazs) [reply ApiClient/Source/Pages/Queries/SearchQuery.swift]: No, unfortunately there is `URLComponents` does not have an initializer like that. Please see:
https://developer.apple.com/documentation/foundation/urlcomponents

File: ApiClient/Source/Pages/Queries/SearchQuery.swift (hunk 4 above)
Comment (vinczebalazs) [reply ApiClient/Source/Pages/Queries/SearchQuery.swift]: `URLComponents(string: host)` is a failable initalizer, thus why it returns an optional.

Comment (dmfs) [reply ApiClient/Source/Pages/Queries/SearchQuery.swift]: That's too bad. Apparently Apple doesn't understand the concept of immutability.

Anyway, why do you continue when creating the URL failed? If the constructor failed you should raise an exception instead of moving on. It means that something went really wrong.

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 5]
```diff
@@ -27,25 +27,31 @@ import Foundation
...
+        urlComponents?.queryItems = queryItems
+        self.url = urlComponents?.url
     }
     
     /// Automatically add locale and location parameter to the pages URL
     convenience init() {
-        self.init(locale: Locale.preferredLanguages[0].components(separatedBy: "-")[0], location: Locale.current.regionCode!)
+        let localeQuery = URLQueryItem(name: "locale", value: Locale.preferredLanguages[0].components(separatedBy: "-")[0])
+        let locationQuery = URLQueryItem(name: "location", value: Locale.current.regionCode!)
+        self.init(path: "", queryItems: [localeQuery,locationQuery])
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: redundant variables …

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 6]
```diff
@@ -27,25 +27,31 @@ import Foundation
...
     }
     
     /// Manualy specify locale and location parameters
-    convenience init(locale: String, location: String){
-        self.init(resource: "pages/?locale=\(locale)?location=\(location)")
+    convenience init(locale: String, location: String) {
+        self.init(path: "", queryItems: [URLQueryItem(name: "locale", value: locale), URLQueryItem(name: "location", value: location)])
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: please use new lines to structure your code, that makes it easier to read (see my previous example)

File: ApiClient/Source/Pages/Queries/SearchQuery.swift (hunk 3 above)
Comment (dmfs) [reply ApiClient/Source/Pages/Queries/SearchQuery.swift]: That's too bad. Apparently Apple doesn't understand the concept of immutability.

File: ApiClient/Source/Pages/Queries/SearchQuery.swift [hunk 7]
```diff
@@ -27,22 +27,25 @@ import Foundation
...
     }
 
     /// Initiliaze with a query string
     convenience init(query: String) {
-        let locale = Locale.preferredLanguages[0].components(separatedBy: "-")[0]
-        let encodedQuery = query.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!
-        self.init(resource: "/pages/search?q=\(encodedQuery)&locale=\(locale)")
+        let localeQuery = URLQueryItem(name: "locale", value: Locale.preferredLanguages[0].components(separatedBy: "-")[0])
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/SearchQuery.swift]: redundant …

File: ApiClient/Source/Query.swift [hunk 8]
```diff
@@ -29,10 +29,12 @@ import Alamofire
...
     
-    var resource: String { get }
+    var url: URL? { get }
+    var host: String { get }
     var method: HTTPMethod { get }
     var parameters: Parameters { get }
     var headers: HTTPHeaders { get }
     
-    func handleResult(with data: Data) -> Result
+    init(path: String, queryItems: [URLQueryItem])
```
Comment (dmfs) [inline ApiClient/Source/Query.swift]: why do you add a constructor to a protocol? The constructor and the parameter it takes are specific to each query and there is no point in forcing it to have any specific constructor.

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 9]
```diff
@@ -26,26 +26,42 @@
...
 
-final class HomePageQuery: Query {
-    typealias Result = JSONPage?
-    
-    let resource: String
+final class HomePageQuery: Query {    
+    typealias Result = JSONPage
+
+    let url: URL
+    let host: String = "https://api.schedjoules.com/pages"
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: that's not a host, it's a full blown URL

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 10]
```diff
@@ -26,26 +26,42 @@
...
     let method: HTTPMethod = .get
     let parameters: Parameters = [:]
     let headers: HTTPHeaders = ["Accept" : "application/json", "Content-Type" : "application/json"]
 
-    required init(resource: String) {
-        self.resource = resource
+    required init?(path: String?, queryItems: [URLQueryItem]) {
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: why do you take a path if all convenience ctors pass `null` anyway and if the path is actually fixed?

File: ApiClient/Source/Pages/Queries/SearchQuery.swift [hunk 11]
```diff
@@ -27,22 +27,35 @@ import Foundation
...
-    let resource: String
+    let url: URL
+    let host: String = "https://api.schedjoules.com/pages"
     let method: HTTPMethod = .get
     let parameters: Parameters = [:]
     let headers: HTTPHeaders = ["Accept" : "application/json", "Content-Type" : "application/json"]
     
-    required init(resource: String) {
-        self.resource = resource
+    required init?(path: String?, queryItems: [URLQueryItem]) {
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/SearchQuery.swift]: same here, why allow passing a custom path?

File: ApiClient/Source/Pages/Queries/SinglePageQuery.swift [hunk 12]
```diff
@@ -26,27 +26,41 @@
...
-    let resource: String
+    let url: URL
+    let host: String = "https://api.schedjoules.com"
     let method: HTTPMethod = .get
     let parameters: Parameters = [:]
     let headers: HTTPHeaders = ["Accept" : "application/json", "Content-Type" : "application/json"]
     
-    required init(resource: String) {
-        self.resource = resource
+    required init?(path: String?, queryItems: [URLQueryItem]) {
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/SinglePageQuery.swift]: why allow passing a path?

File: ApiClient/Tests/ParsingTests.swift [hunk 13]
```diff
@@ -52,8 +52,7 @@ class ApiClientTests: XCTestCase {
         
         let decoder = JSONDecoder()
         do {
-            let pageItem = try decoder.decode(JSONPageItem.self, from: data)
-            print(pageItem)
+            _ = try decoder.decode(JSONPageItem.self, from: data)
```
Comment (dmfs) [inline ApiClient/Tests/ParsingTests.swift]: How are these changes related to the String to URL change?

File: ApiClient/Source/Query.swift [hunk 14]
```diff
@@ -29,10 +29,11 @@ import Alamofire
 protocol Query {
     associatedtype Result
     
-    var resource: String { get }
+    var url: URL { get }
+    var host: String { get }
```
Comment (dmfs) [inline ApiClient/Source/Query.swift]: why does the protocol still need the host when we have a url?

File: ApiClient/Source/Pages/Queries/SearchQuery.swift [hunk 15]
```diff
@@ -27,22 +27,34 @@ import Foundation
...
+        self.url = urlComponents.url!
     }
 
     /// Initiliaze with a query string
-    convenience init(query: String) {
-        let locale = Locale.preferredLanguages[0].components(separatedBy: "-")[0]
-        let encodedQuery = query.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!
-        self.init(resource: "/pages/search?q=\(encodedQuery)&locale=\(locale)")
+    convenience init?(query: String) {
+        let localeQuery = URLQueryItem(name: "locale", value: Locale.preferredLanguages[0].components(separatedBy: "-")[0])
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/SearchQuery.swift]: this local variable can be inlined

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 16]
```diff
@@ -26,26 +26,41 @@
...
+        let localeQuery = URLQueryItem(name: "locale", value: Locale.preferredLanguages[0].components(separatedBy: "-")[0])
+        let locationQuery = URLQueryItem(name: "location", value: Locale.current.regionCode!)
+        self.init(queryItems: [localeQuery,locationQuery])
     }
     
     /// Manualy specify locale and location parameters
-    convenience init(locale: String, location: String){
-        self.init(resource: "pages/?locale=\(locale)?location=\(location)")
+    convenience init?(locale: String, location: String) {
+        self.init(queryItems: [URLQueryItem(name: "locale", value: locale), URLQueryItem(name: "location", value: location)])
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: can you reformat this to put the two `URLQueryItem`s onto separate lines? This would help the reader.

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 17]
```diff
@@ -26,26 +26,41 @@
...
+            return nil
+        }
+        self.url = urlComponents.url!
     }
     
     /// Automatically add locale and location parameter to the pages URL
-    convenience init() {
-        self.init(locale: Locale.preferredLanguages[0].components(separatedBy: "-")[0], location: Locale.current.regionCode!)
+    convenience init?() {
+        let localeQuery = URLQueryItem(name: "locale", value: Locale.preferredLanguages[0].components(separatedBy: "-")[0])
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: redundant local variables can be inlined

File: ApiClient/Source/Query.swift (hunk 14 above)
Comment (vinczebalazs) [reply ApiClient/Source/Query.swift]: not needed, removing.

File: ApiClient/Tests/ParsingTests.swift (hunk 13 above)
Comment (vinczebalazs) [reply ApiClient/Tests/ParsingTests.swift]: As well as improving test coverage I cleaned up the code in the tests.

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 18]
```diff
@@ -26,26 +26,33 @@
 import Foundation
 import Alamofire
 
-final class HomePageQuery: Query {
-    typealias Result = JSONPage?
-    
-    let resource: String
+final class HomePageQuery: Query {
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: remove whitespace at the end of the line

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 19]
```diff
@@ -26,26 +26,33 @@
...
+        self.init(queryItems: [
+                URLQueryItem(name: "locale", value: Locale.preferredLanguages[0].components(separatedBy: "-")[0]),
+                URLQueryItem(name: "location", value: Locale.current.regionCode!)])
     }
     
     /// Manualy specify locale and location parameters
-    convenience init(locale: String, location: String){
-        self.init(resource: "pages/?locale=\(locale)?location=\(location)")
+    convenience init(locale: String, location: String) {
+        self.init(queryItems: [URLQueryItem(name: "locale", value: locale), URLQueryItem(name: "location", value: location)])
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: code style:
```swift
self.init(queryItems: [
    URLQueryItem(name: "locale", value: locale),
    URLQueryItem(name: "location", value: location)])
```

File: ApiClient/Source/Pages/Queries/SearchQuery.swift [hunk 20]
```diff
@@ -27,22 +27,26 @@ import Foundation
...
+        // Set the url property to the url constructed from the components
+        self.url = urlComponents!.url!
     }
 
     /// Initiliaze with a query string
     convenience init(query: String) {
-        let locale = Locale.preferredLanguages[0].components(separatedBy: "-")[0]
-        let encodedQuery = query.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed)!
-        self.init(resource: "/pages/search?q=\(encodedQuery)&locale=\(locale)")
+        self.init(queryItems: [URLQueryItem(name: "search?q", value: query),
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/SearchQuery.swift]: `search` is not a query item, it's part of the path, isn't it?

File: ApiClient/Source/Pages/Queries/SinglePageQuery.swift [hunk 21]
```diff
@@ -26,27 +26,34 @@
...
+        // Set the url property to the url constructed from the components
+        self.url = urlComponents!.url!
     }
     
-    /// Initialize with a given Page ID and automatically add locale and
-    /// location parameter to the pages URL
+    /// Initialize with a given Page ID and automatically add locale parameter
     convenience init(pageID: String) {
-        self.init(pageID: pageID, locale: Locale.preferredLanguages[0].components(separatedBy: "-")[0])
+        self.init(pageID: pageID,
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/SinglePageQuery.swift]: why don't you just delegate to the other ctor which takes a `locale` argument? Avoid code duplication.

File: ApiClient/Tests/ParsingTests.swift [hunk 22]
```diff
@@ -1,6 +1,6 @@
 //
-//  ApiClientTests.swift
-//  ApiClientTests
+//  ParsingTests.swift
+//  ParsingTests
```
Comment (dmfs) [inline ApiClient/Tests/ParsingTests.swift]: How many times do we have to repeat `ParsingTests` until a Swift developer understands it? ;-)

File: ApiClient/Source/QueryError.swift [hunk 23]
```diff
@@ -0,0 +1,14 @@
+//
+//  QueryError.swift
+//  ApiClient
+//
+//  Created by Balazs Vincze on 2018. 04. 11..
+//  Copyright © 2018. SchedJoules. All rights reserved.
+//
```
Comment (dmfs) [inline ApiClient/Source/QueryError.swift]: License header? Did you add it to your template?

File: ApiClient/Source/QueryError.swift [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+//  QueryError.swift
+//  ApiClient
```
Comment (dmfs) [inline ApiClient/Source/QueryError.swift]: ApiClient or QueryError? Which one is it?

File: ApiClient/Source/Pages/Queries/SinglePageQuery.swift [hunk 25]
```diff
@@ -27,26 +27,32 @@ import Foundation
...
+    let host: String = "https://api.schedjoules.com"
     let method: HTTPMethod = .get
     let parameters: Parameters = [:]
     let headers: HTTPHeaders = ["Accept" : "application/json", "Content-Type" : "application/json"]
     
-    required init(resource: String) {
-        self.resource = resource
+    private init(pageID: String, queryItems: [URLQueryItem]) {
+        // Initialize url components from a string
+        var urlComponents = URLComponents(string: "https://api.schedjoules.com/pages/\(pageID)")
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/SinglePageQuery.swift]: will this automatically urlencode pageID?

File: ApiClient/Source/Pages/Queries/HomePageQuery.swift [hunk 26]
```diff
@@ -27,25 +27,33 @@ import Foundation
...
+        // Add query items to the url
+        urlComponents!.queryItems = queryItems
+        // Set the url property to the url constructed from the components
+        self.url = urlComponents!.url!
     }
     
     /// Automatically add locale and location parameter to the pages URL
     convenience init() {
-        self.init(locale: Locale.preferredLanguages[0].components(separatedBy: "-")[0], location: Locale.current.regionCode!)
+        self.init(queryItems: [
```
Comment (dmfs) [inline ApiClient/Source/Pages/Queries/HomePageQuery.swift]: better call the other ctor to avoid code duplication

File: ApiClient/Source/Pages/Queries/SinglePageQuery.swift (hunk 25 above)
Comment (vinczebalazs) [reply ApiClient/Source/Pages/Queries/SinglePageQuery.swift]: PageID is numeric, does it require encoding?

File: ApiClient/Source/SchedJoulesApiClient.swift [hunk 27]
```diff
@@ -61,17 +57,20 @@ final class SchedJoulesApiClient: Api {
         for (key,value) in query.headers {
             headers[key] = value
         }
-        
+
```
Comment (dmfs) [inline ApiClient/Source/SchedJoulesApiClient.swift]: spaces

File: ApiClient/Source/Pages/Queries/SinglePageQuery.swift (hunk 25 above)
Comment (dmfs) [reply ApiClient/Source/Pages/Queries/SinglePageQuery.swift]: This ctor takes a string, so the user could pass anything. This ctor should still work when we switch to string IDs.
